### PR TITLE
[inductor] Add reduction loop peeling for Triton codegen

### PR DESCRIPTION
**Pull request by liqiangxl**
Add loop peeling optimization for non-persistent reductions that splits the reduction loop into an unmasked main loop (vectorizable) and a masked tail loop, gated behind config.triton.loop_peeling (default off).

During the single masked codegen pass, load() and reduction() register masked_line→unmasked_line mappings in a dictionary. The peeled loop emitter uses this map to derive unmasked lines without a second codegen pass, making the optimization zero-cost at compile time.

Fix issue https://github.com/pytorch/pytorch/issues/148402
Fixes #122840
Idea was proposed by @shunting314 

**Result:** Performance of bf16 softmax (32768, 50257) on GB200: 
(a) 1.42x speedup for static shape
(b) 1.22x speedup for dynamic shape (no vectorization due to lack of stride divisible by 16 hint)

See updated [design doc](https://github.com/liqiangxl/my_design_docs/blob/main/inductor/loop_peeling.md) for details.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -5242,68 +5344,13 @@ def codegen_body(self):
...
-                        self.body.writeline(
-                            DeferredLine(
-                                self.block_ptr_to_buffer[block_ptr],
-                                f"{block_ptr} = tl.advance({block_ptr}, {V.kernel.index_to_str(advancement)})",
-                            )
-                        )
-
-                # Invalidate any cache entries that came from inside the loop.
-                self.cse.invalidate(self.outside_loop_vars)
-                tree.cache_clear()
```
Comment (liqiangxl) [inline torch/_inductor/codegen/triton.py:5306]: This whole `else` branch is moved to function `_emit_reduction_loops`

File: test/inductor/test_codegen_triton.py [hunk 2]
```diff
@@ -258,6 +258,110 @@ def test_signature_of_fp8_dtypes(self):
...
+    def _check_peeling(self, fn, *inputs):
+        compiled = torch.compile(fn)
+        result, codes = run_and_get_code(compiled, *inputs)
+        self.assertIn("r0_numel_aligned", "\n".join(codes))
+        expected = fn(*inputs)
+        if isinstance(expected, tuple):
+            for r, e in zip(result, expected):
+                torch.testing.assert_close(r, e)
+        else:
+            torch.testing.assert_close(result, expected)
```
Comment (shunting314) [inline test/inductor/test_codegen_triton.py]: torch.testing.assert_close can handle list/tuple itself

File: torch/_inductor/codegen/simd.py [hunk 3]
```diff
@@ -2051,6 +2051,15 @@ def codegen_node_schedule_with_kernel(self, node_schedule, kernel):
 
             kernel.finalize_indexing(all_indexing.keys())
 
+            # Enable loop peeling map if eligible — the codegen pass will
+            # populate masked→unmasked line mappings as a side effect.
+            loop_trees = [t for t in kernel.range_trees if t.is_loop]
+            if (
+                hasattr(kernel, '_should_peel_reduction_loop')
```
Comment (shunting314) [inline torch/_inductor/codegen/simd.py]: Add _should_peel_reduction_loop to base class and return false for backend that does not support loop peeling.

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -5242,68 +5344,13 @@ def codegen_body(self):
...
+            peeling = (
+                self._unmasked_line_map is not None
+                and not self.pointer_advancements.get(loop_trees[0].symt)
```
Comment (shunting314) [inline torch/_inductor/codegen/triton.py]: With block ptr, self._unmasked_line_map should already be None?

**Review by shunting314 — COMMENTED** [review]
Any reason not turning it on by default?

Can you also share some benchmarking results in the PR summary?

**Comment by shunting314** [comment]
Add @eellison to take another look.

File: torch/_inductor/codegen/simd.py (hunk 3 above)
Comment (liqiangxl) [reply torch/_inductor/codegen/simd.py]: Good point! Added to `SIMDKernel`, default is `false`

**Review by eellison — COMMENTED** [review]
I'm a little worried about how hard it is to reason about some of these transformations in our current stack, and feel confident about their correctness as well as stacking & modularity of transformations. 

We are looking at a way of making it easier to do IR transformations after initial codegen. I am going to clean up a bit and share. 

Having said, certainly this refactor wouldn't be ready imminently so I'm not sure whether we should defer this PR or not.

File: torch/_inductor/codegen/triton.py (hunk 4 above)
Comment (liqiangxl) [reply torch/_inductor/codegen/triton.py]: yes, `_should_peel_reduction_loop` ensures block ptr is not used.

**Comment by liqiangxl** [comment]
> Any reason not turning it on by default?

Changed to enabled by default. I initially wanted to let experienced/internal users run more tests before turning it on, but that may have been too cautious. Enabling it by default should allow us to better leverage CI coverage.

> Can you also share some benchmarking results in the PR summary?

Added. Consistent with your original observation in the issue, we see about a 1.4× speedup for the static case.
For the dynamically compiled case, there is also a 1.2× speedup, although vectorized loads are still not generated due to a symbolic stride.

**Comment by liqiangxl** [comment]
> I'm a little worried about how hard it is to reason about some of these transformations in our current stack, and feel confident about their correctness as well as stacking & modularity of transformations.
> 
> We are looking at a way of making it easier to do IR transformations after initial codegen. I am going to clean up a bit and share.
> 
> Having said, certainly this refactor wouldn't be ready imminently so I'm not sure whether we should defer this PR or not.

I believe `IR transformations` should make a lot of optimizations easier and compose well when you stack multiple of them. Happy to migrate to the IR transformation framework when it's ready. The current peeling logic is contained in `_unmasked_line_map` should be straightforward to swap out. In the meantime this gives us the perf win today for both static and dynamic kernels.

**Comment by liqiangxl** [comment]
> We are looking at a way of making it easier to do IR transformations after initial codegen. I am going to clean up a bit and share.

Hi @eellison, will this new way of doing IR transformations making [persistent reduction optimization #179711](https://github.com/pytorch/pytorch/issues/179711) easier? Or in other words, would it be better to wait for this new approach before working on #179711?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by liqiangxl** [comment]
Current status:
Simplified to emit a main loop over `r0_numel_aligned` with `r0_mask = True`, then a tail loop with the normal dynamic mask. It does not use the older `clone-pass` (has compile overhead) or `string-map` (fragile) approach.

First loop, over aligned range without bounday check (`r0_mask = True`). 
```python
    r0_numel_aligned = (r0_numel // R0_BLOCK) * R0_BLOCK
    for r0_offset in tl.range(0, r0_numel_aligned, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = True
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp0 = tl.load(in_ptr0 + (r0_1 + 1027*x0), r0_mask & xmask, eviction_policy='evict_first', other=0.0)
```
Second loop, over the tailing elements with regular boundary check
```
    for r0_offset in tl.range(r0_numel_aligned, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_1 = r0_index
        tmp0 = tl.load(in_ptr0 + (r0_1 + 1027*x0), r0_mask & xmask, eviction_policy='evict_first', other=0.0)
```
